File: version.old/pp.first/cpp/Machine.cpp

 /**
 *  
 *  This file is the implementation of a virtual machine which
    has some ideas drawn from the sed stream editor tool. 
    More documentation for the machine and auxiliary programs
    is available at http://bumble.sf.net/machine

    The basic purpose of the set of class which use the Machine 
    class is allow the creation of a language interpreter which
    will interpret a parsing-language.
    
    The language looks like this
    
    pop;pop;pop;
    /stringg*equals*value/ 
    {
      clear;
      add "text";
      push;
    }

    The code has been compiled with the bloodshed devcc compiler.
 *
 *
 *  @author http://bumble.sf.net
 */

#include <iostream>
#include <fstream>
#include <vector>
#include <stack>
#include <string>
#include <sstream>
#include <ctype.h>
#include "Machine.h" 

using namespace std;
 

  //--------------------------------------------
  Machine::Machine()
  {
    stack<string> tokenstack;
    Tape tape;
    char peep = 'a';
    string influx("");
    string workarea("");
    string lastOperation("NEW MACHINE");
    bool stackChanged = false;
  }

  //--------------------------------------------
  /* looks if the stack flag has been set. */
  bool Machine::stackFlag()
  {
    return this->stackChanged;
  } //-- 

  //--------------------------------------------
  /* sets the stack changed flag to true. This is useful for multiple
   * reductions of the stack in one cycle */
  void Machine::setFlag()
  {
    this->stackChanged = true;
  } //-- 

  //--------------------------------------------
  /* sets the stack changed flag to true. This is useful for multiple
   * reductions of the stack in one cycle */
  void Machine::resetFlag()
  {
    this->stackChanged = false;
  } //-- 

//--------------------------------------------
  string Machine::workspace()
  {
    return this->workarea;
  } //-- 

  //--------------------------------------------
  void Machine::peek()
  {
    this->workarea += this->peep;  
  } //-- 

  //--------------------------------------------
  bool Machine::peek(string sTest)
  {
      if ((sTest == "[:digit:]") && (!isdigit(this->peep)))
      {
        return true;
      } 

      if ((sTest == "[:letter:]") && (!isalpha(this->peep)))
      {
        return true;
      } 

      if ((sTest == "[:space:]") && (!isspace(this->peep)))
      {
        return true;
      } 

          
    return false;
  } //-- 

//--------------------------------------------
  /* reads one char from the input stream and update the machine */
  void Machine::readNext(istream& inputstream)
  {
    stringstream ss;
    ss << this->peep;
    this->workarea.append(ss.str());          
    inputstream.get(this->peep);
    this->lastOperation.clear();
    this->lastOperation.append("read next");
    
  }  

  //--------------------------------------------
  /* reads while the peep is */
  void Machine::readWhile(istream& inputstream, string sWhile)
  {          
    while(!inputstream.eof())
    {
      if ((sWhile.length() == 1) && (this->peep != sWhile.at(0)))
      {
        this->lastOperation.clear();
        this->lastOperation.append("read while not");
        return;
      } 

      if ((sWhile == "[:digit:]") && (!isdigit(this->peep)))
      {
        this->lastOperation.clear();
        this->lastOperation.append("read while");
        return;
      } 

      if ((sWhile == "[:letter:]") && (!isalpha(this->peep)))
      {
        this->lastOperation.clear();
        this->lastOperation.append("read while");
        return;
      }

      if ((sWhile == "[:space:]") && (!isspace(this->peep)))
      {
        this->lastOperation.clear();
        this->lastOperation.append("read while");
        return;
      } 


      this->workarea += this->peep;          
      inputstream.get(this->peep);
    }

    this->lastOperation.clear();
    this->lastOperation.append("read while");
    
  }  //-- method: readWhile

  //--------------------------------------------
  /* reads while the peep is not */
  void Machine::readWhileNot(istream& inputstream, string sWhileNot)
  {          
    while(!inputstream.eof())
    {
      if ((sWhileNot.length() == 1) && (this->peep == sWhileNot.at(0)))
      {
        this->lastOperation.clear();
        this->lastOperation.append("read while not");
        return;
      } 

      if ((sWhileNot == "[:digit:]") && (isdigit(this->peep)))
      {
        this->lastOperation.clear();
        this->lastOperation.append("read while not");
        return;
      } 

      if ((sWhileNot == "[:letter:]") && (isalpha(this->peep)))
      {
        this->lastOperation.clear();
        this->lastOperation.append("read while not");
        return;
      } 

      if ((sWhileNot == "[:space:]") && (isspace(this->peep)))
      {
        this->lastOperation.clear();
        this->lastOperation.append("read while not");
        return;
      } 

      this->workarea += this->peep;          
      inputstream.get(this->peep);
    }

    this->lastOperation.clear();
    this->lastOperation.append("read while not");
    
  }  //-- method: readWhileNot

//--------------------------------------------
  /* reads until the accumulator end with the string */
  void Machine::readUntil(istream& inputstream, string sTerminator)
  {          
    while(!inputstream.eof())
    {                        
      this->workarea += this->peep;          
      inputstream.get(this->peep);
      if (this->endsWith(sTerminator))
      {
        this->lastOperation.clear();
        this->lastOperation.append("read until");
        return;
      } 
    }

    this->lastOperation.clear();
    this->lastOperation.append("read until");
    
  }  

  //--------------------------------------------
  /* reads until the accumulator end with the string */
  void Machine::readUntil(istream& inputstream, string sTerminator, string sNotTerminator)
  {          

    while(!inputstream.eof())
    {                        
      this->workarea += this->peep;          
      inputstream.get(this->peep);
      if (this->endsWith(sTerminator) && (!this->endsWith(sNotTerminator)))
      {
        this->lastOperation.clear();
        this->lastOperation.append("read until but not");
        return;
      } 
    }
    this->lastOperation.clear();
    this->lastOperation.append("read until but not");
    
  }  

  //--------------------------------------------
  bool Machine::endsWith(string sFinal)
  {          

    if (sFinal.length() > this->workarea.length())
     { return false; }

    if (this->workarea.substr(this->workarea.length() - sFinal.length()) == sFinal)
    {
      return true;
    }
    return false;
  }   

  /* si el espai de treball comenca amb la cadena */
  //--------------------------------------------
  bool Machine::beginsWith(string sBegin)
  {          
    if (sBegin == "")
     { return false; } 
	  
    if (sBegin.length() > this->workarea.length())
     { return false; }

    if (this->workarea.substr(0, sBegin.length()) == sBegin)
     { return true; }

    return false;
  } 

  //--------------------------------------------
  /** this method may not be necessary, can check with pops */
  int Machine::stacksize()
  {
    return this->tokenstack.size();
  }  

  //--------------------------------------------
  bool Machine::matches(string sTest)
  {
    if (this->workarea == sTest)
      { return true; }

    return false;
  }  

  //--------------------------------------------
  /* determines if the workspace is a space character */
  bool Machine::isSpace()
  {
    
    for (int ii = 0; ii < this->workarea.length(); ii++)
    {
      if (!isspace(this->workarea.at(ii)))
       { return false; }
    }
    return true;
  } //-- 

  //--------------------------------------------
  /* determines if the workspace is a digit character */
  bool Machine::isDigit()
  {
    if (this->workarea.length() != 1)
    {
      return false;
    }            

    if (isdigit(this->workarea.at(0)))
      { return true; }

    return false;
  } //-- 

  //--------------------------------------------
  /* determines if the workspace is a letter character */
  bool Machine::isLetter()
  {
    if (this->workarea.length() != 1)
    {
      return false;
    }            

    if (isalpha(this->workarea.at(0)))
      { return true; }

    return false;
  } //-- 

  //--------------------------------------------
  /*
  bool Machine::isUnicode()
  {
    if (this->workarea.length() != 1)
    {
      return false;
    }            

    if (Character.isDefined(this->workarea.at(0)))
      { return true; }

    return false;
  } //-- 
  */

  //--------------------------------------------
  /* to allow simple pattern testing for literal values */
  bool Machine::workspaceInRange(char cStart, char cEnd)
  {
    if (this->workarea.length() > 1)
     { return false;}

    char cCharacter = this->workarea.at(0);

    if (cCharacter < cStart)
     { return false; }

    if (cCharacter > cEnd)
     { return false; }

    return true;
  }

  //--------------------------------------------
  /* to allow simple pattern testing for literal values */
  bool Machine::matches(char cStart, char cEnd)
  {
    if (this->workarea.length() > 1)
     { return false;}

    char cCharacter = this->workarea.at(0);

    if (cCharacter < cStart)
     { return false; }

    if (cCharacter > cEnd)
     { return false; }

    return true;
  }


  //--

  //--------------------------------------------
  /** decrements the pointer to the tape by one */
  void Machine::decrementTape()
  {
    this->tape.decrementPointer();
    this->lastOperation.clear();
    this->lastOperation.append("decrement-tape");
  } //-- 

  //--------------------------------------------
  /** increments the pointer to the tape by one */
  void Machine::incrementTape()
  {
    this->tape.incrementPointer();
    this->lastOperation.clear();
    this->lastOperation.append("increment-tape");
  } //-- 


  //--------------------------------------------
  /** puts the workspace into the current item of the tape.
   *  The workspace is not changed  */
  void Machine::put()
  {
    this->tape.put(this->workarea);
    this->lastOperation.clear();
    this->lastOperation.append("put");
  } //-- 

  //--------------------------------------------
  /** gets the current item of the tape and adds it to
   *  the end of the workspace */
  void Machine::get()
  {
    this->workarea.append(this->tape.get());
    this->lastOperation.clear();
    this->lastOperation.append("get");
  } //-- 

  //--------------------------------------------
  /** inserts the last item of the stack at the front of the workspace,
   * adding a bar character "|" to separate the token. If the
   * stack is empty, there is no change to the machine.
   */
  void Machine::pop()
  {
    if (this->tokenstack.empty())
    {
      this->lastOperation.clear();
      this->lastOperation.append("pop [stack empty]");
      return;
    }

    string sSymbol(this->tokenstack.at(this->tokenstack.size() - 1));
    this->tokenstack.pop_back();


    //-- the automatic adding of a bar character during a pop operation
    //-- is not desirable, it is better to let the script writer put
    //-- in the token delimiter character
    //--string sText(this->replaceText(sSymbol, "|", "&bar;"));
    this->workarea.insert(0, sText);
    this->tape.decrementPointer();
    this->lastOperation.clear();
    this->lastOperation.append("pop");
  } //-- 

  string Machine::replaceText(string sOriginal, string sReplace, string sReplacement)
  {
    if (sOriginal.length() == 0)
     { return ""; }
    
    string::size_type iFirst = sOriginal.find(sReplace, 0 );


    if (iFirst == string::npos)
    {
      return sOriginal;
    }
        sOriginal.replace(iFirst, sReplace.length(), sReplacement);  
    return sOriginal;
  } //-- 

  //--------------------------------------------
  /** replaces all occurences of a string in the workarea 
   */ 
  void Machine::replace(string sReplace, string sReplacement)
  {
    if (this->workarea.length() == 0)
     { return; }

    string::size_type iLocation = this->workarea.find(sReplace, 0);
    if (iLocation == string::npos)
     { return; }

    string sOriginal("");
    sOriginal = this->workarea;
    this->workarea.clear();

    for (int ii = 0; ii < sOriginal.length(); ii++)
    {
      this->workarea.append(sOriginal.substr(ii, 1));

      if (this->endsWith(sReplace)) 
      {
        this->workarea.erase(this->workarea.length() - sReplace.length());
	this->workarea.append(sReplacement);
      }
    } //-- for
    return;
  } //-- method: replace

  //--------------------------------------------
  /** pushes the contents of the workspace onto the stack.
   *  The first token on the workspace is deleted.
   */ 
  void Machine::push()
  {

    if (this->workarea.length() < 1)
    {
      this->lastOperation.clear();
      this->lastOperation.append("push [workspace empty]");
      return;
    }

    this->lastOperation.clear();
    this->lastOperation.append("push");

    string::size_type iFirstStar = this->workarea.find("*", 0 );

    if (iFirstStar == string::npos)
    {
      //string sText(this->replaceText(this->workarea, "&bar;", "|"));
      this->tokenstack.push_back(this->workarea);
      this->tape.incrementPointer();
      this->workarea.clear();
      return;
    }

    string sFirstToken(this->workarea.substr(0, iFirstStar + 1));
    //string sText(this->replaceText(sFirstToken, "&bar;", "|"));
    this->tokenstack.push_back(sFirstToken);
    this->workarea = this->workarea.substr(iFirstStar + 1);
    this->tape.incrementPointer();
    return;
  } //-- method: push

  //--------------------------------------------
  /** prints the workspace to standard out */
  void Machine::print()
  {
    cout << this->workarea;    
    this->lastOperation.clear();
    this->lastOperation.append("print");
  } 

  //--------------------------------------------
  /** adds a piece of text to the end of the workspace */
  void Machine::add(string sText)
  {
    this->workarea.append(sText);
    this->lastOperation.clear();
    this->lastOperation.append("add");

  } 

  //--------------------------------------------
  /** clears the workspace */
  void Machine::clear()
  {
    this->workarea.clear();
    this->lastOperation.clear();
    this->lastOperation.append("clear");
  } 

  //--------------------------------------------
  /**  anyade una nueva linea al espacio */
  void Machine::newline()
  {
    this->workarea.append("\n");
    this->lastOperation.clear();
    this->lastOperation.append("newline");
  } 

  //--------------------------------------------
  /** indents each line of the workspace, which may
   *  be useful for print code fragments */
  void Machine::indent()
  {
    string sText(this->workarea);
    this->workarea.clear();
    char cCurrent;

    this->workarea.append("  ");
    for (int ii = 0; ii < sText.length(); ii++)
    {
      cCurrent = sText[ii];
      this->workarea += cCurrent;
      if (cCurrent == '\n')
       { this->workarea.append("  "); }

    } //-- for
    this->lastOperation.clear();
    this->lastOperation.append("indent");

  } //-- method: indent 

//--------------------------------------------
/* removes one character from the end of the workspace */
  void Machine::clip()
  {          

    if (this->workarea.length() == 0)
    { return; }

    this->workarea.erase(this->workarea.length() - 1, 1);
  }   
  //--------------------------------------------
  /**  */
  string Machine::toString()
  {
    return this->printState();
  } //-- method:

  //--------------------------------------------
  /** returns a description of the current state of the machine, displays the 
   *  contents of the tokenstack, tape and the workspace.
   */ 
  string Machine::printState()
  {
    string sMessage("");
    sMessage.append("last operation:" + this->lastOperation);
    sMessage.append("\n");
    sMessage.append("PEEP     :[");
    sMessage += this->peep;
    sMessage.append("]");
    sMessage.append("\n");
    sMessage.append("WORKSPACE:[");
    sMessage.append(this->workarea);
    sMessage.append("]");
    sMessage.append("\n");
    sMessage.append("STACKFLAG:[");
    if (this->stackChanged)
     { sMessage.append("true"); }
    else
     { sMessage.append("false"); }
    sMessage.append("]");
    sMessage.append("\n");

    sMessage.append("STACK    :");
    sMessage.append(this->printStack());
    sMessage.append("\n");
    sMessage.append("TAPE     :");
    //sMessage.append("\n");
    sMessage.append(this->tape.print());

    sMessage.append("\n");
    return sMessage;
  } //-- method:

  //--------------------------------------------
  /* shows the contents of the stack in a concise form */ 
  string Machine::printStack()
  {
    string sCurrentItem("");

    string sMessage("[");
    for (int ii = 0; ii < this->tokenstack.size(); ii++)
    {
      sMessage.append(this->tokenstack.at(ii)); 
      if (ii < this->tokenstack.size() - 1)
        { sMessage.append(", "); }
    }
    
    sMessage.append("]");
    return sMessage;
  } //-- method:

  

File: version.old/pp.first/java/try1/machine/Machine.java

package machine;

import java.net.*;
import java.io.*;
import java.util.*;
import machine.Tape;

/**
 *  
 *  Is a virtual machine for parsing. It has some ideas
 *  drawn from the sed tool.
 *
 *  @author http://bumble.sf.net
 */
 
  
public class Machine extends Object
{
  //--------------------------------------------
  private static String NEWLINE = System.getProperty("line.separator");
  //--------------------------------------------
  private Stack stack;
  //--------------------------------------------
  private Tape tape;
  //--------------------------------------------
  private StringBuffer workspace;
  //--------------------------------------------
  private StringBuffer lastOperation;
  //--------------------------------------------
  /** todo: */
  private Integer peep;

  //--------------------------------------------
  public Machine()
  {
    this.stack = new Stack();
    this.tape = new Tape();
    this.workspace = new StringBuffer();
    this.lastOperation = new StringBuffer("NEW MACHINE");
  }


  //--------------------------------------------
  public String getWorkspace()
  {
    return this.workspace.toString();
  }  

  //--------------------------------------------
  public String workspace()
  {
    return this.workspace.toString();
  }  

  //--------------------------------------------
  /** this method may not be necessary, can check with pops */
  public int stacksize()
  {
    return this.stack.size();
  } //-- 

  //--------------------------------------------
  public boolean matches(String sTest)
  {
    if (this.workspace.toString().equals(sTest))
      { return true; }

    return false;
  } //-- 

  //--------------------------------------------
  /* determines if the workspace is a space character */
  public boolean isSpace()
  {
    if (this.workspace.toString().length() != 1)
    {
      return false;
    }	    

    if (Character.isSpace(this.workspace.toString().charAt(0)))
      { return true; }

    return false;
  } //-- 

  //--------------------------------------------
  /* determines if the workspace is a digit character */
  public boolean isDigit()
  {
    if (this.workspace.toString().length() != 1)
    {
      return false;
    }	    

    if (Character.isDigit(this.workspace.toString().charAt(0)))
      { return true; }

    return false;
  } //-- 

  //--------------------------------------------
  /* determines if the workspace is a letter character */
  public boolean isLetter()
  {
    if (this.workspace.toString().length() != 1)
    {
      return false;
    }	    

    if (Character.isLetter(this.workspace.toString().charAt(0)))
      { return true; }

    return false;
  } //-- 

  //--------------------------------------------
  public boolean isUnicode()
  {
    if (this.workspace.toString().length() != 1)
    {
      return false;
    }	    

    if (Character.isDefined(this.workspace.toString().charAt(0)))
      { return true; }

    return false;
  } //-- 

  //--------------------------------------------
  /* to allow simple pattern testing for literal values */
  public boolean workspaceInRange(char cStart, char cEnd)
  {
    if (this.workspace.length() > 1)
     { return false;}

    char cCharacter = this.workspace.toString().charAt(0);

    if (cCharacter < cStart)
     { return false; }

    if (cCharacter > cEnd)
     { return false; }

    return true;
  }

  //--------------------------------------------
  /* to allow simple pattern testing for literal values */
  public boolean matches(char cStart, char cEnd)
  {
    if (this.workspace.length() > 1)
     { return false;}

    char cCharacter = this.workspace.toString().charAt(0);

    if (cCharacter < cStart)
     { return false; }

    if (cCharacter > cEnd)
     { return false; }

    return true;
  }

  //--------------------------------------------
  /** decrements the pointer to the tape by one */
  public void decrementTape()
  {
    this.tape.decrementPointer();
    this.lastOperation.setLength(0);
    this.lastOperation.append("decrement-tape");
  } //-- 

  //--------------------------------------------
  /** increments the pointer to the tape by one */
  public void incrementTape()
  {
    this.tape.incrementPointer();
    this.lastOperation.setLength(0);
    this.lastOperation.append("increment-tape");
  } //-- 

  //--------------------------------------------
  /** puts the workspace into the current item of the tape.
   *  The workspace is not changed  */
  public void put()
  {
    this.tape.put(this.workspace.toString());
    this.lastOperation.setLength(0);
    this.lastOperation.append("put");
  } //-- 

  //--------------------------------------------
  /** gets the current item of the tape and adds it to
   *  the end of the workspace */
  public void get()
  {
    this.workspace.append(this.tape.get());
    this.lastOperation.setLength(0);
    this.lastOperation.append("get");
  } //-- 

  //--------------------------------------------
  /** inserts the last item of the stack at the front of the workspace,
   * adding a bar character "|" to separate the token. If the
   * stack is empty, there is no change to the machine.
   */
  public void pop()
  {
    if (this.stack.empty())
    {
      this.lastOperation.setLength(0);
      this.lastOperation.append("pop [stack empty]");
      return;
    }

    String sSymbol = new String((String)this.stack.pop());
    String sText = new String(ReplaceText.replace(sSymbol, '|', "&bar;"));
    this.workspace.insert(0, sText + "|");
    this.tape.decrementPointer();
    this.lastOperation.setLength(0);
    this.lastOperation.append("pop");
  } //-- 

  //--------------------------------------------
  /** pushes the contents of the workspace onto the stack.
   *  The first token on the workspace is deleted.
   */ 
  public void push()
  {

    if (this.workspace.toString().length() < 1)
    {
      this.lastOperation.setLength(0);
      this.lastOperation.append("push [workspace empty]");
      return;
    }

    this.lastOperation.setLength(0);
    this.lastOperation.append("push");

    int iFirstBar = this.workspace.toString().indexOf('|');

    if (iFirstBar < 0)
    {
      String sText = new String(
        ReplaceText.replaceString(this.workspace.toString(), "&bar;", "|"));
      this.stack.push(sText);
      this.tape.incrementPointer();
      this.workspace.setLength(0);
      return;
    }

    String sFirstToken = new String(
      this.workspace.toString().substring(0, iFirstBar));
    String sText = new String(ReplaceText.replaceString(sFirstToken, "&bar;", "|"));
    this.stack.push(sText);
    this.workspace = new StringBuffer(
      this.workspace.toString().substring(iFirstBar + 1));
    this.tape.incrementPointer();

  } //-- method: push


  //--------------------------------------------
  /** prints the workspace to standard out */
  public void print()
  {
    System.out.print(this.workspace.toString());
    this.lastOperation.setLength(0);
    this.lastOperation.append("print");
  } 

  //--------------------------------------------
  /** adds a piece of text to the end of the workspace */
  public void add(String sText)
  {
    this.workspace.append(sText);
    this.lastOperation.setLength(0);
    this.lastOperation.append("add");

  } 

  //--------------------------------------------
  /** clears the workspace */
  public void clear()
  {
    this.workspace.setLength(0);
    this.lastOperation.setLength(0);
    this.lastOperation.append("clear");
  } 

  //--------------------------------------------
  /**  anyade una nueva linea al espacio */
  public void newline()
  {
    this.workspace.append(NEWLINE);
    this.lastOperation.setLength(0);
    this.lastOperation.append("newline");
  } 

  //--------------------------------------------
  /** todo: read the input stream until the workspace ends with the text */
  public void until(String sText)
  {
  }

  public void until(String sText, String sNotText)
  {}

  /** reads the input stream until the workspace ends with the 
   *  text in the current element of the tape */
  public void until()
  {}

  //--------------------------------------------
  /** todo: */
  public void while()
  {
  }

  
  //--------------------------------------------
  /** todo: */
  public void whileNot()
  {
  }

  /** todo: */
  public void crash()
  {
  }

  //--------------------------------------------
  /** indents each line of the workspace, which may
   *  be useful for print code fragments */
  public void indent()
  {
    String sText = new String(this.workspace.toString());
    this.workspace.setLength(0);
    char cCurrent;

    this.workspace.append("  ");
    for (int ii = 0; ii < sText.length(); ii++)
    {
      cCurrent = sText.charAt(ii);
      this.workspace.append(cCurrent);
      if (this.workspace.toString().endsWith(NEWLINE))
       { this.workspace.append("  "); }

    } //-- for
    this.lastOperation.setLength(0);
    this.lastOperation.append("indent");

  } //-- method: indent 

  //--------------------------------------------
  /**  */
  public String toString()
  {
    return "";
  } //-- method:

  //--------------------------------------------
  /** returns a description of the current state of the machine, displays the 
   *  contents of the stack, tape and the workspace.
   */ 
  public String printState()
  {
    StringBuffer sbMessage = new StringBuffer();
    sbMessage.append(NEWLINE);
    sbMessage.append("last operation:" + this.lastOperation);
    sbMessage.append(NEWLINE);
    sbMessage.append("WORKSPACE:[");
    sbMessage.append(this.workspace.toString());
    sbMessage.append("]");
    sbMessage.append(NEWLINE);

    sbMessage.append("STACK    :");
    sbMessage.append(this.stack.toString());
    sbMessage.append(NEWLINE);
    sbMessage.append("TAPE     :");
    sbMessage.append(NEWLINE);
    sbMessage.append(this.tape.print());

    sbMessage.append(NEWLINE);
    return sbMessage.toString();
  } //-- method:


  //--------------------------------------------
  /** provides a command loop to test the machine operations and view the machine */
  public static void main(String[] args) throws Exception
  {
    
    StringBuffer sbUsageMessage = new StringBuffer("");
    sbUsageMessage.append("test usage: java Machine ");
    sbUsageMessage.append(NEWLINE);

    StringBuffer sbMessage = new StringBuffer("");


    if (args.length > 2)
    {	    
      System.out.println(sbUsageMessage);
      System.exit(-1);
    }

    //String sText = args[0];
    //char cChar = args[1].charAt(0);


    Machine testMachine = new Machine();

    BufferedReader brUserInput = new BufferedReader(
        new InputStreamReader(System.in));

     String sCommand = "";
     // SoundDownload sdCurrent = (SoundDownload)ee.nextElement();

     StringBuffer sbUserMessage = new StringBuffer("");
     sbUserMessage.setLength(0);

     sbUserMessage.append("Commands;");
     sbUserMessage.append(NEWLINE);
     sbUserMessage.append(" a - add to the workspace");
     sbUserMessage.append(NEWLINE);
     sbUserMessage.append(" c - clear the work space [clear, cl]");
     sbUserMessage.append(NEWLINE);
     sbUserMessage.append(" i - indents the work space [indent]");
     sbUserMessage.append(NEWLINE);
     sbUserMessage.append(" n - add newline to the work space [newline, nl]");
     sbUserMessage.append(NEWLINE);
     sbUserMessage.append(" p - push the workspace onto the stack [push]");
     sbUserMessage.append(NEWLINE);
     sbUserMessage.append(" o - pop the stack into the workspace [pop]");
     sbUserMessage.append(NEWLINE);
     sbUserMessage.append(" u - put the workspace on the tape [put]");
     sbUserMessage.append(NEWLINE);
     sbUserMessage.append(" g - get the tape item into the workspace ");
     sbUserMessage.append(NEWLINE);
     sbUserMessage.append(" - - decrement the tape pointer");
     sbUserMessage.append(NEWLINE);
     sbUserMessage.append(" + - increment the tape pointer");
     sbUserMessage.append(NEWLINE);
     sbUserMessage.append(" h - help, show this message [?]");
     sbUserMessage.append(NEWLINE);
     sbUserMessage.append(" q - quit");
     sbUserMessage.append(NEWLINE);
     System.out.println(sbUserMessage);

     //----------------------------------
     //-- the command loop
     //--
     while (!sCommand.equals("q"))
     {
       //--------------------------------
       // 
       if (sCommand.startsWith("a"))
       {
         testMachine.add(sCommand.substring(1));               
       }

       //--------------------------------
       // 
       if (sCommand.equals("c") || sCommand.equals("clear"))
       {
         testMachine.clear();
       }

       //--------------------------------
       // 
       if (sCommand.equals("n") || sCommand.equals("newline"))
       {
         testMachine.newline();
       }

       //--------------------------------
       // 
       if (sCommand.equals("i") || sCommand.equals("indent"))
       {
         testMachine.indent();
       }

       //--------------------------------
       // 
       if (sCommand.equals("p") || sCommand.equals("push"))
       {
         testMachine.push();               
       }

       //--------------------------------
       // 
       if (sCommand.equals("o") || sCommand.equals("pop"))
       {
         testMachine.pop();               
       }

       //--------------------------------
       // 
       if (sCommand.equals("u") || sCommand.equals("put"))
       {
         testMachine.put();
       }

       //--------------------------------
       // 
       if (sCommand.equals("g") || sCommand.equals("get"))
       {
         testMachine.get();
       }

       //--------------------------------
       // 
       if (sCommand.equals("-"))
       {
         testMachine.decrementTape();               
       }

       //--------------------------------
       // 
       if (sCommand.equals("+"))
       {
         testMachine.incrementTape();               
       }

       //--------------------------------
       // 
       if (sCommand.equals("?") || sCommand.equals("h"))
       {
         System.out.println(sbUserMessage);               
       }


       System.out.print(testMachine.printState());
       System.out.print(">");
       sCommand = brUserInput.readLine();
     } //-- while


    System.out.println(testMachine.printState());
    

  } //-- main()
  
} //-- Machine class
